src/app/state/wallet/saga.ts

```diff
@@ -114,29 +107,30 @@ export function* openWalletFromMnemonic() {
         path: account.path,
         privateKey: account.privateKey,
         publicKey: account.publicKey,
-        selectImmediately: false,
         type: account.type,
+        selectImmediately: account === accounts[0], // Select first
       }),
     )
   }
-  const existingWallet = yield* call(getWalletByAddress, accounts[0].address)
-  yield* put(walletActions.selectWallet(existingWallet ? existingWallet.id : newWalletId))
 }
 
 /**
  * Adds a wallet to the existing wallets
- * If the wallet exists already, do nothingg
+ * If the wallet exists already, do nothing
  * If it has "selectImmediately", we select it immediately
  */
-export function* addWallet({ payload: newWallet }: PayloadAction<AddWalletPayload>) {
+export function* addWallet({ payload }: PayloadAction<AddWalletPayload>) {
+  const { selectImmediately, ...newWallet } = payload
   const existingWallet = yield* call(getWalletByAddress, newWallet.address)
   if (!existingWallet) {
     yield* put(walletActions.walletOpened(newWallet))
   }
 
   const walletId = existingWallet ? existingWallet.id : newWallet.id
 
-  if (newWallet.selectImmediately) {
+  if (selectImmediately) {
+    yield* put(walletActions.selectWallet(undefined)) // Workaround so useRouteRedirects detects selecting the same account
+    yield* delay(1)
```

Comment:
ouch, can you add a comment that this is used to avoid React batch updates?

Reply:
yep yep

Tho I'm not sure why this would be batched in react 17 :shrug: